File: day/12/part/1/speed.js
let moons = [
  {x: -13, y: -13, z: -13, vx: 0, vy: 0, vz: 0, id: '0'},
  {x: 5, y: -8, z: 3, vx: 0, vy: 0, vz: 0, id: '1'},
  {x: -6, y: -10, z: -3, vx: 0, vy: 0, vz: 0, id: '2'},
  {x: 0, y: 5, z: -5, vx: 0, vy: 0, vz: 0, id: '3'},
]

function toStr(obj) {
  let str = []
  obj.forEach(ob => {
    ['x', 'y', 'z', 'vx', 'vy', 'vz'].forEach(c => {
      str.push(ob[c])
    })
  })
  return str.join(',')
}

let orgStr = []
for (let i = 0; i < moons.length; i++) {
  orgStr.push(moons[i].x, moons[i].y, moons[i].z, moons[i].vx, moons[i].vy, moons[i].vz)
}
orgStr = orgStr.join(',')

function getPairs(arr) {
  return arr.map( (v, i) => arr.slice(i + 1).map(w => [v, w]) ).flat()
}

let pairs = getPairs(moons)

console.log(pairs)

function step() {
  
}


let objs = new Set()
let steps = 0
console.time('time')
while (true) {
  steps += 1
  
  for (let i = 0; i < pairs.length; i++) {
    let pair = pairs[i]

    if (pair[0].x === pair[1].x) {
        
    } else if (pair[0].x > pair[1].x) {
      pair[0].vx -= 1
      pair[1].vx += 1
    } else if (pair[0].x < pair[1].x) {
      pair[0].vx += 1
      pair[1].vx -= 1
    }

    if (pair[0].y === pair[1].y) {
        
    } else if (pair[0].y > pair[1].y) {
      pair[0].vy -= 1
      pair[1].vy += 1
    } else if (pair[0].y < pair[1].y) {
      pair[0].vy += 1
      pair[1].vy -= 1
    }

    if (pair[0].z === pair[1].z) {
        
    } else if (pair[0].z > pair[1].z) {
      pair[0].vz -= 1
      pair[1].vz += 1
    } else if (pair[0].z < pair[1].z) {
      pair[0].vz += 1
      pair[1].vz -= 1
    }
  }

  for (let i = 0; i < moons.length; i++) {
    moons[i].x += moons[i].vx
    moons[i].y += moons[i].vy
    moons[i].z += moons[i].vz

  }

  let str = []
  for (let i = 0; i < moons.length; i++) {
    str.push(moons[i].x, moons[i].y, moons[i].z, moons[i].vx, moons[i].vy, moons[i].vz)
  }
  str = str.join(',')

  if (str === orgStr) {
    console.log(str)
    break
  }

  if (steps % 1e6 === 0) {
    console.timeLog('time', steps)
  }
  // if (time % 100 === 0) {
  //   console.log(moons.map((v, i) => `${i}: ${v.vx}, ${v.vy}, ${v.vz},`))
  // }
}

// let total = 0

// moons.forEach(moon => {
//   total += (Math.abs(moon.x) + Math.abs(moon.y) + Math.abs(moon.z)) * (Math.abs(moon.vx) + Math.abs(moon.vy) + Math.abs(moon.vz))
// })

console.log(objs)
console.log(steps)
console.timeEnd('time')

File: day/12/part/1/speed.c
#include <stdio.h>

void main() {
  int omoons[4][6] = {
    {-8, -10, 0, 0, 0, 0},
    { 5, 5, 10, 0, 0, 0},
    { 2, -7, 3, 0, 0, 0},
    { 9, -8, -3, 0, 0, 0}
  };

  int moons[4][6] = {
    {-8, -10, 0, 0, 0, 0},
    { 5, 5, 10, 0, 0, 0},
    { 2, -7, 3, 0, 0, 0},
    { 9, -8, -3, 0, 0, 0}
  };

  // int omoons[4][6] = {
  //   { -13, -13, -13, 0, 0, 0 },
  //   { 5, -8, -3, 0, 0, 0 },
  //   { -6, -10, -3, 0, 0, 0 },
  //   { 0, 5, -5, 0, 0, 0 },
  // };

  // int moons[4][6] = {
  //   { -13, -13, -13, 0, 0, 0 },
  //   { 5, -8, -3, 0, 0, 0 },
  //   { -6, -10, -3, 0, 0, 0 },
  //   { 0, 5, -5, 0, 0, 0 },
  // };

  int *pairs[6][2] = {
    { moons[0], moons[1] },
    { moons[0], moons[2] },
    { moons[0], moons[3] },
    { moons[1], moons[2] },
    { moons[1], moons[3] },
    { moons[2], moons[3] }
  };
  int steps = 0;

  while (1)
  {
    for (int i = 0; i < 6; i++) {
      for (int c = 0; c < 3; c++) {
        if ( (pairs[i][0])[c] > (pairs[i][1])[c]) {
          (pairs[i][0])[c + 3]++;
          (pairs[i][1])[c + 3]--;
        }
        if ( (pairs[i][0])[c] < (pairs[i][1])[c]) {
          (pairs[i][0])[c + 3]--;
          (pairs[i][1])[c + 3]++;
        }
      }
    }

    int notFound = 0;

    for (int i = 0; i < 4; i++) {
      moons[i][0] = moons[i][0] + moons[i][3];
      moons[i][1] = moons[i][1] + moons[i][4];
      moons[i][2] = moons[i][2] + moons[i][5];

      if (moons[i][0] != omoons[i][0]
        || moons[i][1] != omoons[i][1]
        || moons[i][2] != omoons[i][2]
        || moons[i][3] != omoons[i][3]
        || moons[i][4] != omoons[i][4]
        || moons[i][5] != omoons[i][5]) {
          notFound = 1;
      }
    }
    
    if (steps == 0) {
      notFound = 1;
    }

    if (!notFound) {
      printf("%d\n", steps);
      break;
    }

    if (steps % 1000000 == 0) {
      printf("%d\n", steps);
    }

    steps++;
  }

  
  
}